chain/txValidator.go

```diff
@@ -874,27 +885,47 @@ func (bvs *BlockValidationState) CleanSubmittedTransactions(txns []*transaction.
 				bvs.subscriptionCount[topic]++
 			}
 		case pb.GENERATE_ID_TYPE:
-			generateIdPayload := payload.(*pb.GenerateID)
-			amount = common.Fixed64(generateIdPayload.RegistrationFee)
-			publicKey := hex.EncodeToString(generateIdPayload.PublicKey)
+			generateIDPayload := payload.(*pb.GenerateID)
+			amount = common.Fixed64(generateIDPayload.RegistrationFee)
+			publicKey := hex.EncodeToString(generateIDPayload.PublicKey)
 			delete(bvs.generateIDs, publicKey)
 		case pb.NANO_PAY_TYPE:
 			npPayload := payload.(*pb.NanoPay)
 			key := nanoPay{hex.EncodeToString(npPayload.Sender), hex.EncodeToString(npPayload.Recipient), npPayload.Id}
 			delete(bvs.nanoPays, key)
+			delete(bvs.nanoPayPayloads, txn.Hash())
 		case pb.ISSUE_ASSET_TYPE:
 		}
 
 		if amount > 0 || fee > 0 {
 			if _, ok := bvs.totalAmount[sender]; ok && bvs.totalAmount[sender] >= amount+fee {
 				bvs.totalAmount[sender] -= amount + fee
-
 				if bvs.totalAmount[sender] == 0 {
 					delete(bvs.totalAmount, sender)
 				}
 			} else {
-				return errors.New("[CleanSubmittedTransactions] inconsistent block validation state")
+				return errors.New("inconsistent block validation state")
+			}
+		}
+	}
+
+	if len(bvs.nanoPayAmount) > 0 {
+		bvs.nanoPayAmount = make(map[common.Uint160]common.Fixed64, len(bvs.nanoPayAmount))
+		for _, npPayload := range bvs.nanoPayPayloads {
+			channelBalance, _, err := DefaultLedger.Store.GetNanoPay(
+				common.BytesToUint160(npPayload.Sender),
+				common.BytesToUint160(npPayload.Recipient),
+				npPayload.Id,
+			)
+			if err != nil {
+				return err
+			}
+
+			if common.Fixed64(npPayload.Amount) <= channelBalance {
+				return errors.New("inconsistent block validation state")
 			}
+
+			bvs.nanoPayAmount[common.BytesToUint160(npPayload.Sender)] += common.Fixed64(npPayload.Amount) - channelBalance
```

Comment:
when does the value of `bvs.nanoPayAmount[common.BytesToUint160(npPayload.Sender)]` set?

Reply:
1. on each new nanopay txn added to txpool
2. on every txn (not just nanopay) removal from txpool